### PR TITLE
Feature/resource_info_metric

### PR DESCRIPTION
**Pull request by rolandvarga**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module github.com/RobustPerception/azure_metrics_exporter
```
Comment (vincentcreusotfx) [inline go.mod]: Is this file needed ???

File: utils.go [hunk 2]
```diff
@@ -41,6 +41,27 @@ func CreateResourceLabels(resourceID string) map[string]string {
 	return labels
 }
 
+func CreateAllResourceLabelsFrom(rm resourceMeta) map[string]string {
+	labels := make(map[string]string)
+	split := strings.Split(rm.ResourceURL, "/")
+	labels["resource_group"] = split[4]
```
Comment (vincentcreusotfx) [inline utils.go]: would you able to put constants for this numbers, and add some comment ?

File: azure.go [hunk 3]
```diff
@@ -270,11 +283,40 @@ func (ac *AzureClient) listByTag(tagName string, tagValue string) ([]string, err
 		return nil, fmt.Errorf("Error unmarshalling response body: %v", err)
 	}
 
-	resources := extractResourceNames(data, subscription)
-
+	var resources []AzureResource
```
Comment (vincentcreusotfx) [inline azure.go]: So you removed the extractResourceName method ? Do you think we could refactor that with a method ?

File: main.go [hunk 4]
```diff
@@ -69,40 +70,47 @@ func (c *Collector) extractMetrics(ch chan<- prometheus.Metric, resource resourc
...
 				prometheus.NewDesc(metricName+"_max", metricName+"_max", nil, labels),
 				prometheus.GaugeValue,
 				metricValue.Maximum,
 			)
 		}
 	}
+
+	infoLabels := CreateAllResourceLabelsFrom(rm)
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc("azure_resource_info", "Azure tags available for resource", nil, infoLabels),
```
Comment (louisfelix) [inline main.go]:
```suggestion
		prometheus.NewDesc("azure_resource_info", "Azure information available for resource", nil, infoLabels),
```

File: utils.go [hunk 5]
```diff
@@ -41,6 +41,27 @@ func CreateResourceLabels(resourceID string) map[string]string {
...
+	labels["resource_type"] = rm.Resource.ResourceType
+
+	for k, v := range rm.Resource.Tags {
+		k = strings.ToLower(k)
+		k = strings.Replace(k, " ", "_", -1)
+		k = strings.Replace(k, "-", "_", -1)
+		k = strings.Replace(k, "/", "_per_", -1)
+		labels[k] = v
+	}
+
```
Comment (louisfelix) [inline utils.go:66]: While we are at it, I would add all available info (ID, Location, Subscription, ManagedBy)

Comment (rolandvarga) [reply utils.go:66]: done!

File: main.go (hunk 4 above)
Comment (rolandvarga) [reply main.go]: updated

File: azure.go (hunk 3 above)
Comment (rolandvarga) [reply azure.go]: I eventually added a method to extend the `AzureResource` type instead

File: utils.go (hunk 2 above)
Comment (rolandvarga) [reply utils.go]: done

File: go.mod (hunk 1 above)
Comment (rolandvarga) [reply go.mod]: The commited vendor directory is still using govendor. I had to update the go prometheus client version to a later version and the most current way to do it is with go modules. This is a proposition to include in the repository for people who are like me and have moved on from using govendor

File: azure.go (hunk 3 above)
Comment (vincentcreusotfx) [reply azure.go]: Nice !

File: go.mod (hunk 1 above)
Comment (vincentcreusotfx) [reply go.mod]: Ok right !

**Review by vincentcreusotfx — APPROVED** [review]
LGTM

**Review by louisfelix — CHANGES_REQUESTED** [review]
It does not work. I have a lot of error when I try to run
1) API version error for `'2019-07-01'`
2) `label dimensions inconsistent with previously collected metrics in the same metric family`

I guess at least one error will be solved if you update the to the correct `client_golang` in `vendors`

**Comment by rolandvarga** [comment]
> It does not work. I have a lot of error when I try to run
> 
> 1. API version error for `'2019-07-01'`
> 2. `label dimensions inconsistent with previously collected metrics in the same metric family`
> 
> I guess at least one error will be solved if you update the to the correct `client_golang` in `vendors`

@louisfelix I committed a `go.mod` file with the updated version of the client. There shouldn't be any errors when you execute using GO modules.

**Comment by louisfelix** [comment]
I don't think it is a good solution. I can currently run it without GO modules, but in this case it uses an outdated `client_golang`. In my case (and in other user's case I am sure), it will "run", but it's broken. This is not intuitive and very error prone; it is a duplication of dependency mechanism that are not in sync. I think you should either propose a complete "breaking" change to GO modules (cleaning vendors all together), or use existing vendor mechanism for your updated `client_golang`.

File: azure.go [hunk 6]
```diff
@@ -70,24 +79,95 @@ type AzureBatchRequestResponse struct {
...
+			log.Println(err)
+			continue
+		}
+		versions = append(versions, APIVersionData{
+			Date:     date,
+			Endpoint: api,
+		})
+	}
+
+	sort.Slice(versions, func(i, j int) bool {
```
Comment (vincentcreusotfx) [inline azure.go]: I like that way but you would'nt need to iterate twice if you just store the last one

File: azure.go [hunk 7]
```diff
@@ -271,12 +339,54 @@ func (ac *AzureClient) listByTag(tagName string, tagValue string, types []string
...
+		return fmt.Errorf("Error unmarshalling response body: %v", err)
+	}
+
+	ac.APIVersions = versionResponse.extractAPIVersions()
+	return nil
+}
+
+func (ac *AzureClient) lookupResourceByID(resourceID string) (AzureResource, error) {
+	resourceType := targetResourceType.FindString(resourceID)
+	apiVersion := ac.APIVersions.findBy(resourceType)
```
Comment (vincentcreusotfx) [inline azure.go:372]: what if apiVersion is not pfound

**Review by vincentcreusotfx — CHANGES_REQUESTED** [review]

File: azure.go [hunk 8]
```diff
@@ -271,12 +339,54 @@ func (ac *AzureClient) listByTag(tagName string, tagValue string, types []string
...
+	if err != nil {
+		return fmt.Errorf("Error unmarshalling response body: %v", err)
+	}
+
+	ac.APIVersions = versionResponse.extractAPIVersions()
+	return nil
+}
+
+func (ac *AzureClient) lookupResourceByID(resourceID string) (AzureResource, error) {
+	resourceType := targetResourceType.FindString(resourceID)
```
Comment (vincentcreusotfx) [inline azure.go:367]: and what if type is empty

Comment (vincentcreusotfx) [reply azure.go:367]:
```suggestion
	resourceType := targetResourceType.FindString(resourceID)
	if resourceType == "" {
		return AzureResource{}, fmt.Errorf("No type found for resource: %s", resourceID)
	}
	apiVersion := ac.APIVersions.findBy(resourceType)

	if apiVersion == "" {
		return AzureResource{}, fmt.Errorf("No api version found for type: %s", resourceType)
	}
```

**Review by vincentcreusotfx — APPROVED** [review]

**Review by louisfelix — APPROVED** [review]
